File: src/bin/0146-lru-cache.py
# @leetup=info id=146 lang=python3 slug=lru-cache



# @leetup=code


import collections


class LRUCache:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache = collections.OrderedDict()

    def get(self, key: int) -> int:
        if key in self.cache:
            self.cache.move_to_end(key)
            return self.cache[key]
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache[key] = value
            self.cache.move_to_end(key)
            return
        if len(self.cache) == self.cap:
            self.cache.popitem(last=False)

        self.cache[key] = value
        self.cache.move_to_end(key)





class Node:
    def __init__(self, key: int, val: int):
        self.key = key
        self.val = val
        self.prev: Node | None = None
        self.next: Node | None = None


class LRUCache_xxxx:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache = {}  # map key to ndoe
        # double linked list
        # Left.next=LRU, right=most recent
        self.left = Node(0, 0)
        self.right = Node(0, 0)
        self.left.next = self.right
        self.right.prev = self.left

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self.remove(node)
            self.insert(node)
            return node.val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.remove(self.cache[key])
        node = Node(key, value)
        self.cache[key] = node
        self.insert(node)
        if len(self.cache) > self.cap:
            # Remove from the list and delete the LRU from the cache
            lru = self.left.next
            self.remove(lru)
            del self.cache[lru.key]

    # For double linked list
    def remove(self, node: Node):
        prev, nxt = node.prev, node.next
        prev.next, nxt.prev = nxt, prev

    def insert(self, node: Node):
        prev, nxt = self.right.prev, self.right
        prev = self.right.prev
        prev.next = node
        nxt.prev = node
        node.next, node.prev = nxt, prev


# @leetup=code

File: src/bin/0146-lru-cache.cpp
// @leetup=info id=146 lang=cpp slug=lru-cache

#include <algorithm>
#include <iostream>
#include <list>
#include <map>
#include <set>
#include <tuple>
#include <unordered_map>
#include <vector>

// @leetup=code

// Ref:
// https://leetcode.com/problems/lru-cache/solutions/931834/c-using-unordered-map-hash-and-list-doubly-linked-list-with-basic-inline-comments/?orderBy=most_votes&languageTags=cpp

// map + linked-list solution.

using key = int;

struct Entry {
  int value;
  std::list<key>::iterator it;
};

class LRUCache {
 public:
  LRUCache(int capacity) : capacity_(capacity) {}

  int get(int key) {
    auto it = cache_.find(key);
    if (it == cache_.end()) {
      return -1;
    }

    // `key` is present. Update `list_`.

    // Remove `key` from `list_`.
    // Move key to front.
    list_.erase(it->second.it);
    list_.push_front(key);
    // Make Entry's `it` point to the front.
    it->second.it = list_.begin();

    return it->second.value;
  }

  void put(int key, int value) {
    auto it = cache_.find(key);
    if (it == cache_.end()) {
      if (cache_.size() == capacity_) {
        // remove the oldest element.
        cache_.erase(list_.back());
        list_.pop_back();
      }

      // push-front key.
      list_.push_front(key);
      cache_[key] = { value, list_.begin()};
    } else {
      // key is present. Move key to the front.
      list_.erase(it->second.it);
      list_.push_front(key);

      // Update Entry.
      it->second.value = value;
      it->second.it = list_.begin();
    }
  }

 private:
  std::unordered_map<key, Entry> cache_;
  // Old keys move to back. New onews enter at front.
  std::list<key> list_;
  size_t capacity_;
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}
